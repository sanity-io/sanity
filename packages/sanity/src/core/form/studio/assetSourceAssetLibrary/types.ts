import {type SanityDocument} from '@sanity/types'

/**
 * Config for the Asset Library Asset Source
 * @alpha
 */
export interface SanityAssetLibraryConfig {
  /**
   * Internal config for the plugin (will be populated by defaults if not explicitly set)
   * @internal
   */
  __internal: {
    /**
     * Wether the plugin is running in a local development environment against local Asset Library app
     */
    isLocalDev: boolean
    /**
     * Wether to use the staging or production asset library environment
     */
    env: 'staging' | 'production'
    /**
     * API version needed for using the Asset Library API
     */
    apiVersion: string
    /**
     * The version of the plugin API to use (routes served by the Asset Library app will be prefixed with this version)
     * @defaultValue 'v1'
     */
    pluginApiVersion: string
    /**
     * Base path for the Asset Library app
     */
    appBasePath: string
    /**
     * Hosts that the plugin uses
     */
    hosts: {
      /**
       * Host for the CDN
       */
      cdn: string
      /**
       * Host for the Asset Library app
       */
      app: string
      /**
       * Host for the Sanity API
       */
      api: string
    }
  }
}

export type AssetDialogAction = {
  type: 'delete' | 'close'
}

export interface AssetMenuAction {
  type: 'delete' | 'showUsage'
}

export type AssetType = 'image' | 'file'

/**
 * Simple version of the Asset Library's Asset type (only stuff that is needed for the plugin)
 * @internal
 */
export interface Asset {
  _id: string
  instanceId: string
  assetType: string
}

/**
 * The type that is returned from the Asset Library for an selected asset item
 * @internal
 */
export interface AssetSelectionItem {
  assetId: string
  assetType: string
  instanceId: string
}

export type PluginPostMessageTokenRequest = {
  type: 'tokenRequest'
}

export type PluginPostMessageTokenResponse = {
  type: 'tokenResponse'
  token: string | null
}

/**
 * Message sent from a plugin page to notify the host that the page is loaded and ready to be interacted with
 */
export type PluginPostMessagePageLoaded = {
  type: 'pageLoaded'
  page: string
}

/**
 * Message sent from a plugin page that a page is unloaded by the user (for closing the dialog and similar)
 */
export type PluginPostMessagePageUnloaded = {
  type: 'pageUnloaded'
  page: string
}

/**
 * Message sent from the plugin that a document has been updated
 */
export type PluginPostMessageDocumentUpdate = {
  type: 'documentUpdate'
  document: SanityDocument | null
}

/**
 * Message sent from the plugin that the user wants to upload files
 */
export type PluginPostMessageUploadFilesRequest = {
  type: 'uploadRequest'
  files: File[]
}

export type PluginPostMessageAssetSelection = {
  type: 'assetSelection'
  selection: AssetSelectionItem[]
}

/**
 * Message sent from the app that the pending uploads are uploaded
 */
export type PluginPostMessageUploadFilesResponse = {
  type: 'uploadResponse'
  assets: Asset[]
}

export type PluginPostMessage =
  | PluginPostMessageTokenRequest
  | PluginPostMessageTokenResponse
  | PluginPostMessageAssetSelection
  | PluginPostMessagePageLoaded
  | PluginPostMessagePageUnloaded
  | PluginPostMessageUploadFilesRequest
  | PluginPostMessageUploadFilesResponse
  | PluginPostMessageDocumentUpdate
