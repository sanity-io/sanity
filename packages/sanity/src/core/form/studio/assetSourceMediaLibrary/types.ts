import {type SanityDocument} from '@sanity/types'

/**
 * Config for the Media Library Asset Source
 * @alpha
 */
export interface SanityMediaLibraryConfig {
  /**
   * Internal config for the plugin (will be populated by defaults if not explicitly set)
   * @internal
   */
  __internal: {
    /**
     * Whether the plugin is running in a local development environment against local Media Library app
     */
    isLocalDev: boolean
    /**
     * Whether to use the staging or production Media Library environment
     */
    env: 'staging' | 'production'
    /**
     * API version needed for using the Media Library API
     */
    apiVersion: string
    /**
     * The version of the plugin API to use (routes served by the Media Library app will be prefixed with this version)
     * @defaultValue 'v1'
     */
    pluginApiVersion: string
    /**
     * Base path for the Media Library app
     */
    appBasePath: string
    /**
     * Hosts that the plugin uses
     */
    hosts: {
      /**
       * Host for the CDN
       */
      cdn: string
      /**
       * Host for the Media Library app
       */
      app: string
      /**
       * Host for the Sanity API
       */
      api: string
    }
  }
}

export type AssetDialogAction = {
  type: 'delete' | 'close'
}

export interface AssetMenuAction {
  type: 'delete' | 'showUsage'
}

export type AssetType = 'image' | 'file'

/**
 * The type that is returned from the Media Library for an selected asset item
 * @internal
 */
export interface AssetSelectionItem {
  asset: {_id: string; _type: string}
  assetInstanceId: string
}

export type PluginPostMessageTokenRequest = {
  type: 'tokenRequest'
}

export type PluginPostMessageTokenResponse = {
  type: 'tokenResponse'
  token: string | null
}

/**
 * Message sent from a plugin page to notify the host that the page is loaded and ready to be interacted with
 */
export type PluginPostMessagePageLoaded = {
  type: 'pageLoaded'
  page: string
}

/**
 * Message sent from a plugin page that a page is unloaded by the user (for closing the dialog and similar)
 */
export type PluginPostMessagePageUnloaded = {
  type: 'pageUnloaded'
  page: string
}

/**
 * Message sent from the plugin that a document has been updated
 */
export type PluginPostMessageDocumentUpdate = {
  type: 'documentUpdate'
  document: SanityDocument | null
}

export type FileStatus = 'pending' | 'uploading' | 'complete' | 'error'

/**
 * Message sent from the plugin when files are uploading
 */
export type PluginPostMessageUploadFilesProgress = {
  type: 'uploadProgress'
  files: {id: string; status: FileStatus; progress: number; error?: Error}[]
}

/**
 * Message sent from the plugin that the user wants to upload files
 */
export type PluginPostMessageUploadFilesRequest = {
  type: 'uploadRequest'
  files: {id: string; file: File}[]
}

export type PluginPostMessageAbortUploadRequest = {
  type: 'abortUploadRequest'
  files?: {id: string}[]
}

export type PluginPostMessageAssetSelection = {
  type: 'assetSelection'
  selection: AssetSelectionItem[]
}

/**
 * Message sent from the app that the pending uploads are uploaded
 */
export type PluginPostMessageUploadFilesResponse = {
  type: 'uploadResponse'
  assets: AssetSelectionItem[]
}

export type PluginPostMessage =
  | PluginPostMessageTokenRequest
  | PluginPostMessageTokenResponse
  | PluginPostMessageAssetSelection
  | PluginPostMessagePageLoaded
  | PluginPostMessagePageUnloaded
  | PluginPostMessageUploadFilesRequest
  | PluginPostMessageAbortUploadRequest
  | PluginPostMessageUploadFilesProgress
  | PluginPostMessageUploadFilesResponse
  | PluginPostMessageDocumentUpdate
