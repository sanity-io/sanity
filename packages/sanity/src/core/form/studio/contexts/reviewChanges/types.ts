/**
 * @internal
 */
export interface ReviewChangesContextValue {
  changesOpen?: boolean
}
