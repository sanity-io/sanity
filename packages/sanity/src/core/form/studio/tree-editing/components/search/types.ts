import {type TreeEditingMenuItem} from '../../types'

export interface SearchableTreeEditingMenuItem extends TreeEditingMenuItem {
  title: string | undefined
  children?: SearchableTreeEditingMenuItem[]
}
