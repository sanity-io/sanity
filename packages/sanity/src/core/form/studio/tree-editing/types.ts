import {type Path, type SchemaType} from 'sanity'

export interface TreeEditingMenuItem {
  children?: TreeEditingMenuItem[]
  parentSchemaType: SchemaType
  path: Path
  schemaType: SchemaType
  value: unknown | undefined
}

export type TreeEditingBreadcrumb = TreeEditingMenuItem
