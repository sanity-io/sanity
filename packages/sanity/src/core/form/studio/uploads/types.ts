import {SanityClient} from '@sanity/client'
import {AssetSourceSpec, SchemaType, AssetMetadataType} from '@sanity/types'
import {Observable} from 'rxjs'
import {FormPatch} from '../../patch'
/**
 *
 * @hidden
 * @beta
 */
export type UploadProgressEvent = {
  type: 'uploadProgress'
  patches: FormPatch[] | null
}

/**
 *
 * @hidden
 * @beta
 */
export type ResolvedUploader = {uploader: Uploader; type: SchemaType}

/**
 *
 * @hidden
 * @beta
 */
export type UploadOptions = {
  metadata?: AssetMetadataType[]
  storeOriginalFilename?: boolean
  label?: string
  title?: string
  description?: string
  creditLine?: string
  source?: AssetSourceSpec
}

/**
 * @internal
 */
export type UploaderDef = {
  type: string
  accepts: string
  upload: (client: SanityClient, file: File, type: SchemaType) => Observable<UploadProgressEvent>
}

/**
 *
 * @hidden
 * @beta
 */
export type Uploader<S extends SchemaType = SchemaType> = {
  type: string
  accepts: string
  upload: (
    client: SanityClient,
    file: File,
    type: S,
    options?: UploadOptions,
  ) => Observable<UploadProgressEvent>
  priority: number
}

/**
 *
 * @hidden
 * @beta
 */
export interface FileLike {
  // mime type
  type: string
  // file name (e.g. somefile.jpg)
  name?: string
}

/**
 *
 * @hidden
 * @beta
 */
export type UploaderResolver<S extends SchemaType = SchemaType> = (
  type: S,
  file: FileLike,
) => Uploader<S> | null
