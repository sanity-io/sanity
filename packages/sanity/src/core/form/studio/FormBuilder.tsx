/* eslint-disable camelcase */
/* eslint-disable react/jsx-handler-names */

import {ObjectSchemaType, Path, ValidationMarker} from '@sanity/types'
import React, {useCallback, useRef} from 'react'
import {FormPatch, PatchChannel, PatchEvent} from '../patch'
import {ObjectFormNode} from '../store/types/nodes'
import {ObjectInputProps} from '../types'
import {useFormBuilder} from '../useFormBuilder'
import {StateTree} from '../store'
import {EMPTY_ARRAY} from '../../util'
import {FormNodePresence} from '../../presence'
import {DocumentFieldAction} from '../../config'
import {FormProvider} from './FormProvider'
import {useFormCallbacks} from './contexts/FormCallbacks'

/**
 * @alpha
 */
export interface FormBuilderProps
  extends Omit<ObjectFormNode, 'level' | 'path' | 'presence' | 'validation' | '_allMembers'> {
  /** @internal */
  __internal_fieldActions?: DocumentFieldAction[]
  /** @internal Considered internal â€“ do not use. */
  __internal_patchChannel: PatchChannel

  autoFocus?: boolean
  changesOpen?: boolean
  collapsedFieldSets: StateTree<boolean> | undefined
  collapsedPaths: StateTree<boolean> | undefined
  focusPath: Path
  focused: boolean | undefined
  id: string
  onChange: (changeEvent: PatchEvent) => void
  onPathBlur: (path: Path) => void
  onPathFocus: (path: Path) => void
  onPathOpen: (path: Path) => void
  onFieldGroupSelect: (path: Path, groupName: string) => void
  onSetFieldSetCollapsed: (path: Path, collapsed: boolean) => void
  onSetPathCollapsed: (path: Path, collapsed: boolean) => void
  presence: FormNodePresence[]
  readOnly?: boolean
  schemaType: ObjectSchemaType
  validation: ValidationMarker[]
  value: {[field in string]: unknown} | undefined
}

/**
 * @alpha
 */
export function FormBuilder(props: FormBuilderProps) {
  const {
    __internal_fieldActions: fieldActions,
    __internal_patchChannel: patchChannel,
    autoFocus,
    changesOpen,
    collapsedFieldSets,
    collapsedPaths,
    focusPath,
    focused,
    groups,
    id,
    members,
    onChange,
    onPathBlur,
    onPathFocus,
    onPathOpen,
    onFieldGroupSelect,
    onSetFieldSetCollapsed,
    onSetPathCollapsed,
    presence,
    readOnly,
    schemaType,
    validation,
    value,
  } = props

  return (
    <FormProvider
      __internal_fieldActions={fieldActions}
      __internal_patchChannel={patchChannel}
      autoFocus={autoFocus}
      changesOpen={changesOpen}
      collapsedFieldSets={collapsedFieldSets}
      collapsedPaths={collapsedPaths}
      focusPath={focusPath}
      focused={focused}
      groups={groups}
      id={id}
      members={members}
      onChange={onChange}
      onPathBlur={onPathBlur}
      onPathFocus={onPathFocus}
      onPathOpen={onPathOpen}
      onFieldGroupSelect={onFieldGroupSelect}
      onSetPathCollapsed={onSetPathCollapsed}
      onSetFieldSetCollapsed={onSetFieldSetCollapsed}
      presence={presence}
      validation={validation}
      readOnly={readOnly}
      schemaType={schemaType}
      value={value}
    >
      <RootInput />
    </FormProvider>
  )
}

function RootInput() {
  const {
    focusPath,
    focused,
    groups,
    id,
    members,
    readOnly,
    renderAnnotation,
    renderBlock,
    renderField,
    renderInlineBlock,
    renderInput,
    renderItem,
    renderPreview,
    schemaType,
    value,
  } = useFormBuilder()

  const {
    onChange,
    onFieldGroupSelect,
    onPathBlur,
    onPathFocus,
    onPathOpen,
    onSetFieldSetCollapsed,
    onSetPathCollapsed,
  } = useFormCallbacks()

  const handleCollapseField = useCallback(
    (fieldName: string) => onSetPathCollapsed([fieldName], true),
    [onSetPathCollapsed],
  )

  const handleExpandField = useCallback(
    (fieldName: string) => onSetPathCollapsed([fieldName], false),
    [onSetPathCollapsed],
  )

  const handleBlur = useCallback(() => onPathBlur(EMPTY_ARRAY), [onPathBlur])

  const handleFocus = useCallback(() => onPathFocus(EMPTY_ARRAY), [onPathFocus])

  const handleChange = useCallback(
    (patch: FormPatch | FormPatch[] | PatchEvent) => onChange(PatchEvent.from(patch)),
    [onChange],
  )

  const focusRef = useRef(null)

  const handleSelectFieldGroup = useCallback(
    (groupName: string) => onFieldGroupSelect(EMPTY_ARRAY, groupName),
    [onFieldGroupSelect],
  )

  const handleOpenField = useCallback((fieldName: string) => onPathOpen([fieldName]), [onPathOpen])

  const handleCloseField = useCallback(() => onPathOpen([]), [onPathOpen])

  const handleCollapseFieldSet = useCallback(
    (fieldSetName: string) => onSetFieldSetCollapsed([fieldSetName], true),
    [onSetFieldSetCollapsed],
  )

  const handleExpandFieldSet = useCallback(
    (fieldSetName: string) => onSetFieldSetCollapsed([fieldSetName], false),
    [onSetFieldSetCollapsed],
  )

  const rootInputProps: Omit<ObjectInputProps, 'renderDefault'> = {
    focusPath,
    elementProps: {
      ref: focusRef,
      id,
      onBlur: handleBlur,
      onFocus: handleFocus,
      'aria-describedby': undefined, // Root input should not have any aria-describedby
    },
    changed: members.some((m) => m.kind === 'field' && m.field.changed),
    focused,
    groups,
    id,
    level: 0,
    members,
    onChange: handleChange,
    onFieldClose: handleCloseField,
    onFieldCollapse: handleCollapseField,
    onFieldSetCollapse: handleCollapseFieldSet,
    onFieldExpand: handleExpandField,
    onFieldSetExpand: handleExpandFieldSet,
    onPathFocus: onPathFocus,
    onFieldOpen: handleOpenField,
    onFieldGroupSelect: handleSelectFieldGroup,
    path: EMPTY_ARRAY,
    presence: EMPTY_ARRAY,
    readOnly,
    renderAnnotation,
    renderBlock,
    renderField,
    renderInlineBlock,
    renderInput,
    renderItem,
    renderPreview,
    schemaType,
    validation: EMPTY_ARRAY,
    value,
  }

  return <>{renderInput(rootInputProps)}</>
}
