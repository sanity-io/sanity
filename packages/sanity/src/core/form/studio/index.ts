export * from './FormBuilder'
export * from './FormProvider'
export * from './contexts'
export * from './defaults'
export * from './focusManagers'
export * from './inputs'
export * from './uploads'
