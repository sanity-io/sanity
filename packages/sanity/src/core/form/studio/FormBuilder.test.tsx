/* eslint-disable camelcase */

import {SanityClient} from '@sanity/client'
import {defineType, Path} from '@sanity/types'
import {render} from '@testing-library/react'
import React, {useMemo} from 'react'
import {createMockSanityClient} from '../../../../test/mocks/mockSanityClient'
import {createTestProvider} from '../../../../test/testUtils/TestProvider'
import {useWorkspace} from '../../studio'
import {EMPTY_ARRAY} from '../../util'
import {createPatchChannel} from '../patch'
import {useFormState} from '../store/useFormState'
import {FormBuilder, FormBuilderProps} from './FormBuilder'

const schemaTypes = [
  defineType({
    type: 'document',
    name: 'test',
    title: 'Test',
    fields: [
      {
        type: 'string',
        name: 'title',
        title: 'Title',
      },
    ],
  }),
]

describe('FormBuilder', () => {
  it('should render a studio form', async () => {
    const client = createMockSanityClient() as unknown as SanityClient
    const TestProvider = await createTestProvider({
      client,
      config: {
        name: 'default',
        projectId: 'test',
        dataset: 'test',
        schema: {types: schemaTypes},
      },
    })

    const focusPath: Path = []
    const openPath: Path = []
    const value = {_id: 'test', _type: 'test'}

    const onChange = jest.fn()
    const onFieldGroupSelect = jest.fn()
    const onPathBlur = jest.fn()
    const onPathFocus = jest.fn()
    const onPathOpen = jest.fn()
    const onSelectFieldGroup = jest.fn()
    const onSetFieldSetCollapsed = jest.fn()
    const onSetPathCollapsed = jest.fn()

    function TestForm() {
      const {schema} = useWorkspace()
      const schemaType = schema.get('test')

      if (!schemaType) {
        throw new Error('missing schema type')
      }

      if (schemaType.jsonType !== 'object') {
        throw new Error('schema type is not an object')
      }

      const patchChannel = useMemo(() => createPatchChannel(), [])

      const formState = useFormState(schemaType, {
        value,
        comparisonValue: value,
        focusPath,
        collapsedPaths: undefined,
        collapsedFieldSets: undefined,
        fieldGroupState: undefined,
        openPath,
        presence: [],
        validation: [],
      })

      const formBuilderProps: FormBuilderProps = useMemo(
        () => ({
          __internal_patchChannel: patchChannel,
          changesOpen: false,
          changed: false,
          collapsedFieldSets: undefined,
          collapsedPaths: undefined,
          focused: formState?.focused,
          focusPath: formState?.focusPath || EMPTY_ARRAY,
          groups: formState?.groups || EMPTY_ARRAY,
          id: formState?.id || '',
          level: formState?.level || 0,
          members: formState?.members || EMPTY_ARRAY,
          onChange,
          onFieldGroupSelect,
          onPathBlur,
          onPathFocus,
          onPathOpen,
          onSelectFieldGroup,
          onSetFieldSetCollapsed,
          onSetPathCollapsed,
          path: EMPTY_ARRAY,
          presence: EMPTY_ARRAY,
          schemaType: formState?.schemaType || schemaType,
          validation: EMPTY_ARRAY,
          value: formState?.value,
        }),
        [formState, patchChannel, schemaType],
      )

      return <FormBuilder {...formBuilderProps} />
    }

    const result = render(
      <TestProvider>
        <TestForm />
      </TestProvider>,
    )

    const titleField = await result.findByTestId('field-title')

    expect(removeClasses(titleField.outerHTML)).toMatchSnapshot()
  })
})

function removeClasses(html: string) {
  return html.replace(/\s+class=".*?"|\s+data-testid="string-input"/g, '')
}
