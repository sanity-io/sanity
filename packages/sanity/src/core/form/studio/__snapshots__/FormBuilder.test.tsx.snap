// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBuilder should render a studio form 1`] = `"<div data-ui=\\"Stack\\" data-testid=\\"field-title\\" data-level=\\"0\\"><div data-as=\\"div\\" data-ui=\\"Flex\\"><div data-as=\\"div\\" data-ui=\\"Box\\"><div data-ui=\\"Stack\\"><div data-as=\\"div\\" data-ui=\\"Flex\\"><label data-ui=\\"Text\\" for=\\"title\\"><span>Title</span></label></div></div></div></div><div><div><div data-testid=\\"change-bar-wrapper\\"><div data-testid=\\"change-bar__field-wrapper\\"><span data-as=\\"span\\" data-ui=\\"TextInput\\" data-scheme=\\"light\\" data-tone=\\"default\\"><span><input data-as=\\"input\\" data-scheme=\\"light\\" data-tone=\\"default\\" id=\\"title\\" type=\\"text\\" value=\\"\\"><span data-border=\\"\\" data-scheme=\\"light\\" data-tone=\\"default\\"></span></span></span></div></div></div></div></div>"`;
